gui/text.3
.\"	$Csoft: text.3,v 1.24 2005/05/11 15:16:00 vedge Exp $
.\"
.\" Copyright (c) 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
.\" <http://www.csoft.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd April 24, 2003
.Dt TEXT 3
.Os
.ds vT Agar API Reference
.ds oS Agar 1.0
.Sh NAME
.Nm text
.Nd agar font engine interface
.Sh SYNOPSIS
.Bd -literal
#include <agar/gui/text.h>
.Ed
.Sh DESCRIPTION
The
.Nm
interface uses an underlying font engine (such as FreeType) to render text.
.Sh INITIALIZATION
.nr nS 1
.Ft void
.Fn AG_TextParseFontSpec "const char *fontspec"
.Pp
.Ft int
.Fn AG_TextInit "enum text_engine te"
.Pp
.Ft void
.Fn AG_TextDestroy "void"
.nr nS 0
.Pp
The
.Fn AG_TextParseFontSpec
function parses a font specification of the form
.Sq face,size,style
(valid separators include
.Sq :,./ )
and assigns the default font.
This function is typically called prior to
.Fn AG_TextInit ,
to parse font specifications on the command line.
.Pp
The
.Fn AG_TextInit
function initializes the text subsystem using the given font engine:
.Bd -literal
enum text_engine {
	TEXT_ENGINE_FREETYPE,		/* FreeType */
	TEXT_ENGINE_BITMAP,		/* Fixed-size bitmap */
	TEXT_ENGINE_PIXMAP		/* Fixed-size pixmap */
};
.Ed
.Pp
.Fn AG_TextInit
selects the default font style by looking at the following
.Xr AG_Config 3
settings:
.Bl -tag -width "font-engine.default-style "
.It Va font-engine.default-font
This specifies the default font face to use.
With
.Dv TEXT_ENGINE_FREETYPE ,
this is the name of the font file to load.
With
.Dv TEXT_ENGINE_BITMAP
and
.Dv TEXT_ENGINE_PIXMAP ,
this is the name of the font bitmap file.
In both cases, the file should reside in one of the directories specified in
the
.Va font-path
.Xr AG_Config 3
setting.
.It Va font-engine.default-size
The default font size in points.
.It Va font-engine.default-style
The default font style, or flags, specified in hexadecimal format.
The meaning of these flags is specific to the font engine being used.
.El
.Pp
The
.Fn AG_TextDestroy
function frees all resources allocated by the font engine.
.Sh FONT INFORMATION
.nr nS 1
.Ft int
.Fn text_get_font_height "ttf_font *font"
.Pp
.Ft int
.Fn text_get_font_ascent "ttf_font *font"
.Pp
.Ft int
.Fn text_get_font_descent "ttf_font *font"
.Pp
.Ft int
.Fn text_get_font_line_skip "ttf_font *font"
.Pp
.nr nS 0
These functions return information specific to a given font.
.Fn text_get_font_height
returns the total height of the font in pixels.
.Fn text_get_font_ascent
returns the offset from the baseline to the top of the font.
.Fn text_get_font_descent
returns the offset from the baseline to the bottom of the font.
.Fn text_get_font_line_skip
returns the recommended spacing in pixels between lines of text.
.Sh RENDERING
.nr nS 1
.Ft "SDL_Surface *"
.Fn AG_TextRenderUnicode "const char *font_name, int font_size, Uint32 color" "const Uint32 *ucs"
.Pp
.Ft "SDL_Surface *"
.Fn AG_TextRender "const char *font_name, int font_size, Uint32 color" "const char *text"
.Pp
.Ft "SDL_Surface *"
.Fn AG_TextRenderGlyph "const char *font_name, int font_size, Uint32 color" "Uint32 uch"
.Pp
.Ft "void"
.Fn AG_TextPrescaleUnicode "const Uint32 *ucs" "int *w" "int *h"
.Pp
.Ft "void"
.Fn AG_TextPrescale "const char *text" "int *w" "int *h"
.nr nS 0
.Pp
The
.Fn AG_TextRender
function renders UTF-8 text to a newly allocated surface.
.Fn AG_TextRenderUnicode
renders UCS-4 text to a newly allocated surface.
.Fn AG_TextRenderGlyph
renders an UCS-4 encoded Unicode character onto a new surface.
A
.Fa font_name
argument of NULL or a
.Fa font_size
argument of -1 indicates that the defaults should be used.
.Pp
The
.Fn AG_TextPrescale
and
.Fn AG_TextPrescaleUnicode
functions return the expected surface dimensions for the given text.
.Sh MESSAGES AND USER INPUT
.nr nS 1
.Ft "void"
.Fn AG_TextMsg "enum ag_text_msg_title title" "const char *format" "..."
.Pp
.Ft "void"
.Fn AG_TextTmsg "enum ag_text_msg_title title" "Uint32 expire" "const char *format" "..."
.Pp
.Ft "void"
.Fn AG_TextPromptFloat "enum text_float_type type" "double *fp" "double min" "double max" "const AG_Unit *unit" "const char *format" "..."
.nr nS 0
.Pp
The
.Fn AG_TextMsg
function displays a text message window containing the given
.Xr printf 3
formatted string, and an
.Sq OK
button.
The
.Fn AG_TextTmsg
variant displays the message for a period of time given in milliseconds.
.Fa title
is one of the following:
.Bd -literal
enum ag_text_msg_title {
	AG_MSG_ERROR,
	AG_MSG_WARNING,
	AG_MSG_INFO
};
.Ed
.Pp
The
.Fn AG_TextPromptFloat
function displays a window prompting the user to enter a floating-point
value.
The
.Fa fp
argument is a pointer to the variable, while
.Fa min
and
.Fa max
define the acceptable range.
Unless
.Fa unit
is NULL, the argument indicates the unit system to use (as in
.Xr units 3 ) .
.Sh TROUBLETONS
Something more sophisticated is needed to deal with unpredictable text surface
dimensions.
Currently,
.Fn AG_TextPrescale
is used prior to allocating fixed-size surfaces, but this is troublesome to
maintain with translations.
.Pp
.Fn AG_TextRenderGlyph
does not work.
.Sh SEE ALSO
.Xr agar 3 ,
.Xr AG_Config 3 ,
.Xr button 3 ,
.Xr combo 3 ,
.Xr label 3 ,
.Xr radio 3 ,
.Xr spinbutton 3 ,
.Xr statusbar 3 ,
.Xr fspinbutton 3 ,
.Xr mfspinbutton 3 ,
.Xr textbox 3 ,
.Xr titlebar 3 ,
.Xr tlist 3 ,
.Xr widget 3
.Bd -literal
The FreeType project - http://www.freetype.org/
Unicode home page - http://www.unicode.org/
.Ed
.Sh HISTORY
The
.Nm
interface first appeared in Agar 1.0.


trunk/gui/text.c
/*	$Csoft: text.c,v 1.109 2005/10/04 17:34:56 vedge Exp $	*/

/*
 * Copyright (c) 2001, 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
 * <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <config/have_freetype.h>
#include <config/have_opengl.h>

#include <core/core.h>
#include <core/view.h>
#include <core/config.h>

#ifdef HAVE_FREETYPE
#include <core/loaders/ttf.h>
#endif
#include <core/loaders/xcf.h>

#include <gui/window.h>
#include <gui/vbox.h>
#include <gui/box.h>
#include <gui/label.h>
#include <gui/button.h>
#include <gui/fspinbutton.h>
#include <gui/textbox.h>
#include <gui/keycodes.h>
#include <gui/unicode.h>

#include <string.h>
#include <stdarg.h>
#include <errno.h>
#include <ctype.h>

int agTextComposition = 1;		/* Built-in input composition */
int agTextBidi = 0;			/* Bidirectionnal text display */
int agTextFontHeight = 0;		/* Default font height (px) */
int agTextFontAscent = 0;		/* Default font ascent (px) */
int agTextFontDescent = 0;		/* Default font descent (px) */
int agTextFontLineSkip = 0;		/* Default font line skip (px) */
int agTextTabWidth = 40;		/* Tab width (px) */
int agTextBlinkRate = 250;		/* Cursor blink rate (ms) */
int agFreetype = 0;			/* Use Freetype font engine */

#define GLYPH_NBUCKETS 1024

static const char *agTextMsgTitles[] = {
	N_("Error"),
	N_("Warning"),
	N_("Information")
};

AG_Mutex agTextLock = AG_MUTEX_INITIALIZER;
static SLIST_HEAD(ag_fontq, ag_font) fonts = SLIST_HEAD_INITIALIZER(&fonts);
AG_Font *agDefaultFont = NULL;

static struct {
	SLIST_HEAD(, ag_glyph) glyphs;
} agGlyphCache[GLYPH_NBUCKETS+1];

static AG_Timeout text_timeout;		/* Timer for AG_TextTmsg() */

AG_Font *
AG_FetchFont(const char *pname, int psize, int pflags)
{
	char path[MAXPATHLEN];
	char name[AG_FONT_NAME_MAX];
	AG_Font *font;
	int size = (psize >= 0) ? psize : AG_Int(agConfig, "font.size");
	u_int flags = (pflags >= 0) ? pflags : AG_Uint(agConfig, "font.flags");

	if (pname != NULL) {
		strlcpy(name, pname, sizeof(name));
	} else {
		AG_StringCopy(agConfig, "font.face", name, sizeof(name));
	}

	AG_MutexLock(&agTextLock);
	SLIST_FOREACH(font, &fonts, fonts) {
		if (font->size == size &&
		    font->flags == flags &&
		    strcmp(font->name, name) == 0)
			break;
	}
	if (font != NULL)
		goto out;

	font = Malloc(sizeof(AG_Font), M_TEXT);
	strlcpy(font->name, name, sizeof(font->name));
	font->size = size;
	font->flags = flags;
	font->c0 = 0;
	font->c1 = 0;
	font->height = 0;
	font->ascent = 0;
	font->descent = 0;
	font->lineskip = 0;
	
	if (AG_ConfigFile("font-path", name, NULL, path, sizeof(path)) == -1)
		goto fail;

#ifdef HAVE_FREETYPE
	if (agFreetype) {
		int tflags = 0;

		dprintf("<%s>: Vector (%d pts)\n", name, size);
		if ((font->p = AG_TTFOpenFont(path, size)) == NULL) {
			goto fail;
		}
		if (flags & AG_FONT_BOLD)	{ tflags|=TTF_STYLE_BOLD; }
		if (flags & AG_FONT_ITALIC)	{ tflags|=TTF_STYLE_ITALIC; }
		if (flags & AG_FONT_UNDERLINE)	{ tflags|=TTF_STYLE_UNDERLINE; }
		AG_TTFSetFontStyle(font->p, tflags);

		font->type = AG_FONT_VECTOR;
		font->height = AG_TTFHeight(font->p);
		font->ascent = AG_TTFAscent(font->p);
		font->descent = AG_TTFDescent(font->p);
		font->lineskip = AG_TTFLineSkip(font->p);
	} else
#endif
	{
		char *spec;
		char *msig, *c0, *c1, *flags;
		AG_Object *fobj;
		AG_Netbuf *buf;

		if ((buf = AG_NetbufOpen(path, "rb", AG_NETBUF_BIG_ENDIAN))
		    == NULL) {
			goto fail;
		}
		font->p = fobj = AG_ObjectNew(NULL, "font");
		fobj->gfx = AG_GfxNew(fobj);
		if (AG_XCFLoad(buf, 0, fobj->gfx) == -1) {
			goto fail_bmp;
		}
		AG_NetbufClose(buf);

		spec = AG_SPRITE(font->p,0).name;
		msig = strsep(&spec, ":");
		c0 = strsep(&spec, "-");
		c1 = strsep(&spec, "-");
		if (msig == NULL || strcasecmp(msig, "MAP") != 0 ||
		    c0 == NULL || c1 == NULL ||
		    c0[0] == '\0' || c1[0] == '\0') {
			AG_SetError("Invalid bitmap fontspec");
			goto fail_bmp;
		}
		font->c0 = (Uint32)strtol(c0, NULL, 10);
		font->c1 = (Uint32)strtol(c1, NULL, 10);
		if (AGOBJECT(font->p)->gfx->nsprites < (font->c1-font->c0)) {
			AG_SetError("Inconsistent bitmap fontspec");
			goto fail_bmp;
		}

		font->type = AG_FONT_BITMAP;
		font->height = AG_SPRITE(font->p,0).su->h;
		font->ascent = font->height;
		font->descent = 0;
		font->lineskip = font->height+2;
	
		dprintf("<%s>: Bitmap '%c'-'%c'\n", name,
		    (char)font->c0, (char)font->c1);
	}

	SLIST_INSERT_HEAD(&fonts, font, fonts);
out:
	AG_MutexUnlock(&agTextLock);
	return (font);
fail_bmp:
	AG_ObjectDestroy(font->p);
	Free(font->p, M_OBJECT);
fail:
	AG_MutexUnlock(&agTextLock);
	Free(font, M_TEXT);
	return (NULL);
}

static Uint32
expire_tmsg(void *obj, Uint32 ival, void *arg)
{
	AG_Window *win = arg;

	AG_ViewDetach(win);
	return (0);
}

/* Initialize the font engine and configure the default font. */
int
AG_TextInit(void)
{
	int i;

#ifdef HAVE_FREETYPE
	if (AG_Bool(agConfig, "font.freetype")) {
		if (AG_TTFInit() == -1) {
			AG_SetError("AG_TTFInit: %s", SDL_GetError());
			return (-1);
		}
		agFreetype = 1;
		if ((agDefaultFont = AG_FetchFont(NULL, -1, -1)) == NULL)
			fatal("%s", AG_GetError());
	} else
#endif
	{
		agFreetype = 0;
		if ((agDefaultFont = AG_FetchFont(NULL, -1, -1)) == NULL)
			fatal("%s", AG_GetError());
	}
	agTextFontHeight = agDefaultFont->height;
	agTextFontAscent = agDefaultFont->ascent;
	agTextFontDescent = agDefaultFont->descent;
	agTextFontLineSkip = agDefaultFont->lineskip;

	for (i = 0; i < GLYPH_NBUCKETS; i++) {
		SLIST_INIT(&agGlyphCache[i].glyphs);
	}
	return (0);
}

static void
free_glyph(AG_Glyph *gl)
{
	SDL_FreeSurface(gl->su);
#ifdef HAVE_OPENGL
	if (agView->opengl)
		glDeleteTextures(1, &gl->texture);
#endif
	Free(gl, M_TEXT);
}

void
AG_TextDestroy(void)
{
	AG_Font *fon, *nextfon;
#ifdef DEBUG
	int maxbucketsz = 0;
#endif
	int i;
	
	for (i = 0; i < GLYPH_NBUCKETS; i++) {
		AG_Glyph *gl, *ngl;
		int bucketsz = 0;

		for (gl = SLIST_FIRST(&agGlyphCache[i].glyphs);
		     gl != SLIST_END(&agGlyphCache[i].glyphs);
		     gl = ngl) {
			ngl = SLIST_NEXT(gl, glyphs);
			free_glyph(gl);
#ifdef DEBUG
			bucketsz++;
#endif
		}
		SLIST_INIT(&agGlyphCache[i].glyphs);
#ifdef DEBUG
		if (bucketsz > maxbucketsz)
			maxbucketsz = bucketsz;
#endif
	}
	
	for (fon = SLIST_FIRST(&fonts);
	     fon != SLIST_END(&fonts);
	     fon = nextfon) {
		nextfon = SLIST_NEXT(fon, fonts);
#ifdef HAVE_FREETYPE
		if (agFreetype) {
			AG_TTFCloseFont(fon->p);
		} else
#endif
		{
			AG_ObjectDestroy(fon->p);
			Free(fon->p, M_OBJECT);
		}
		Free(fon, M_TEXT);
	}
#ifdef HAVE_FREETYPE
	if (agFreetype)
		AG_TTFDestroy();
#endif
}

static __inline__ u_int
hash_glyph(Uint32 ch)
{
	return (ch % GLYPH_NBUCKETS);
}

/* Lookup/insert a glyph in the glyph cache. */
AG_Glyph *
AG_TextRenderGlyph(const char *fontname, int fontsize, Uint32 color,
    Uint32 ch)
{
	AG_Glyph *gl;
	u_int h;

	h = hash_glyph(ch);
	SLIST_FOREACH(gl, &agGlyphCache[h].glyphs, glyphs) {
		if (fontsize == gl->fontsize &&
		    color == gl->color &&
		    ((fontname == NULL && gl->fontname[0] == '\0') ||
		     (strcmp(fontname, gl->fontname) == 0)) &&
		    ch == gl->ch)
			break;
	}
	if (gl == NULL) {
		Uint32 ucs[2];
		SDL_Color c;

		gl = Malloc(sizeof(AG_Glyph), M_TEXT);
		if (fontname == NULL) {
			gl->fontname[0] = '\0';
		} else {
			strlcpy(gl->fontname, fontname, sizeof(gl->fontname));
		}
		gl->fontsize = fontsize;
		gl->color = color;
		gl->ch = ch;

		SDL_GetRGB(color, agVideoFmt, &c.r, &c.g, &c.b);
		ucs[0] = ch;
		ucs[1] = '\0';
		gl->su = AG_TextRenderUnicode(fontname, fontsize, c, ucs);
#ifdef HAVE_OPENGL
		gl->texture = AG_SurfaceTexture(gl->su, gl->texcoord);
#endif
		gl->nrefs = 1;
		SLIST_INSERT_HEAD(&agGlyphCache[h].glyphs, gl, glyphs);
	} else {
		gl->nrefs++;
	}
	return (gl);
}

void
AG_TextUnusedGlyph(AG_Glyph *gl)
{
	if (--gl->nrefs == 0) {
		u_int h;

		h = hash_glyph(gl->ch);
		SLIST_REMOVE(&agGlyphCache[h].glyphs, gl, ag_glyph, glyphs);
		free_glyph(gl);
	}
}

/* Render UTF-8 text onto a newly allocated transparent surface. */
/* XXX use state variables for font spec */
SDL_Surface *
AG_TextRender(const char *fontname, int fontsize, Uint32 color,
    const char *text)
{
	SDL_Color c;
	Uint32 *ucs;
	SDL_Surface *su;

	ucs = AG_ImportUnicode(AG_UNICODE_FROM_UTF8, text);
	SDL_GetRGB(color, agVideoFmt, &c.r, &c.g, &c.b);
	su = AG_TextRenderUnicode(fontname, fontsize, c, ucs);
	free(ucs);
	return (su);
}

static __inline__ SDL_Surface *
AG_TextBitmapGlyph(AG_Font *font, Uint32 c)
{
	if ((font->flags & AG_FONT_UPPERCASE) &&
	    (isalpha(c) && islower(c))) {
		c = toupper(c);
	}
	if (c < font->c0 || c > font->c1) {
		return (AG_SPRITE(font->p,0).su);
	}
	return (AG_SPRITE(font->p,(c - font->c0 + 1)).su);
}

/* Render an UCS-4 text string onto a newly allocated surface. */
SDL_Surface *
AG_TextRenderUnicode(const char *fontname, int fontsize, SDL_Color cFg,
    const Uint32 *text)
{
	AG_Font *font;

#ifdef HAVE_FREETYPE
	if (agFreetype) {
		if (text == NULL || text[0] == '\0') {
			SDL_Surface *su;
	
			return (SDL_CreateRGBSurface(SDL_SWSURFACE,
			    0, 0, 8,
			    0, 0, 0, 0));
		}
		if ((font = AG_FetchFont(NULL, -1, -1)) == NULL) {
			fatal("%s", AG_GetError());
		}
		return (AG_TTFRenderUnicodeSolid(font->p, text, NULL, cFg));
	} else
#endif /* HAVE_FREETYPE */
	{
		size_t i, text_len;
		SDL_Surface *su0, *gsu, *su;
		SDL_Rect rd;
		u_int w = 0, h = 0;
		Uint32 c;

		/* TODO support for size variants or scaling */
		if ((font = AG_FetchFont(NULL, -1, -1)) == NULL) {
			fatal("%s", AG_GetError());
		}
		su0 = AG_SPRITE(font->p,0).su;

		/* Figure out the required surface dimensions. */
		text_len = AG_UCS4Len(text);
		for (i = 0; i < text_len; i++) {
			c = text[i];
			gsu = AG_TextBitmapGlyph(font, text[i]);
			w += gsu->w;
			h = MAX(h,gsu->h);
		}

		/* Allocate the final surface. */
		su = SDL_CreateRGBSurface(SDL_SWSURFACE,
		    w, h, 32,
		    agSurfaceFmt->Rmask,
		    agSurfaceFmt->Gmask,
		    agSurfaceFmt->Bmask,
		    0);
		if (su == NULL)
			fatal("SDL_CreateRGBSurface: %s", SDL_GetError());

		/* Blit the glyphs. */
		rd.x = 0;
		rd.y = 0;
		for (i = 0; i < text_len; i++) {
			gsu = AG_TextBitmapGlyph(font, text[i]);
			rd.w = gsu->w;
			rd.h = gsu->h;
			SDL_BlitSurface(gsu, NULL, su, &rd);
			rd.x += gsu->w;
		}
		
		SDL_SetColorKey(su, SDL_SRCCOLORKEY|SDL_RLEACCEL,
		    0);
		SDL_SetAlpha(su, su0->flags&(SDL_SRCALPHA|SDL_RLEACCEL),
		    su0->format->alpha);

		return (su);
	}
}

/* Return the expected size of an Unicode text element. */
void
AG_TextPrescaleUnicode(const Uint32 *ucs, int *w, int *h)
{
	SDL_Surface *su;
	SDL_Color c;

	c.r = 0;
	c.g = 0;
	c.b = 0;

	/* TODO use the bounding box in freetype mode */
	su = AG_TextRenderUnicode(NULL, -1, c, ucs);
	if (w != NULL) { *w = (int)su->w; }
	if (h != NULL) { *h = (int)su->h; }
	SDL_FreeSurface(su);
}

/* Return the expected surface size for a UTF-8 string. */
void
AG_TextPrescale(const char *text, int *w, int *h)
{
	Uint32 *ucs;

	ucs = AG_ImportUnicode(AG_UNICODE_FROM_UTF8, text);
	AG_TextPrescaleUnicode(ucs, w, h);
	free(ucs);
}

/* Display a message. */
void
AG_TextMsg(enum ag_text_msg_title title, const char *format, ...)
{
	char msg[AG_LABEL_MAX];
	AG_Window *win;
	AG_VBox *vb;
	va_list args;

	va_start(args, format);
	vsnprintf(msg, sizeof(msg), format, args);
	va_end(args);

	win = AG_WindowNew(AG_WINDOW_NORESIZE|AG_WINDOW_NOCLOSE|
	    AG_WINDOW_NOMINIMIZE|AG_WINDOW_NOMAXIMIZE|AG_WINDOW_NOBORDERS);
	AG_WindowSetCaption(win, "%s", _(agTextMsgTitles[title]));
	AG_WindowSetPosition(win, AG_WINDOW_CENTER, 1);

	vb = AG_VBoxNew(win, 0);
	AG_LabelNew(vb, AG_LABEL_STATIC, msg);

	vb = AG_VBoxNew(win, AG_VBOX_HOMOGENOUS|AG_VBOX_WFILL|AG_VBOX_HFILL);
	AG_ButtonAct(vb, _("Ok"), AG_BUTTON_FOCUS, AGWINDETACH(win));

	AG_WindowShow(win);
}

/* Display a message for a given period of time. */
void
AG_TextTmsg(enum ag_text_msg_title title, Uint32 expire, const char *format,
    ...)
{
	char msg[AG_LABEL_MAX];
	AG_Window *win;
	AG_VBox *vb;
	va_list args;

	va_start(args, format);
	vsnprintf(msg, sizeof(msg), format, args);
	va_end(args);

	win = AG_WindowNew(AG_WINDOW_NORESIZE|AG_WINDOW_NOCLOSE|
	    AG_WINDOW_NOMINIMIZE|AG_WINDOW_NOMAXIMIZE|AG_WINDOW_NOBORDERS);
	AG_WindowSetCaption(win, "%s", _(agTextMsgTitles[title]));
	AG_WindowSetPosition(win, AG_WINDOW_CENTER, 1);

	vb = AG_VBoxNew(win, 0);
	AG_LabelNew(vb, AG_LABEL_STATIC, msg);
	AG_WindowShow(win);

	AG_LockTimeouts(NULL);
	if (AG_TimeoutIsScheduled(NULL, &text_timeout)) {
		AG_ViewDetach((AG_Window *)text_timeout.arg);
		AG_DelTimeout(NULL, &text_timeout);
	}
	AG_UnlockTimeouts(NULL);

	AG_SetTimeout(&text_timeout, expire_tmsg, win, AG_TIMEOUT_LOADABLE);
	AG_AddTimeout(NULL, &text_timeout, expire);
}

/* Prompt the user with a choice of options. */
AG_Window *
AG_TextPromptOptions(AG_Button **bOpts, u_int nbOpts, const char *fmt, ...)
{
	char text[AG_LABEL_MAX];
	AG_Window *win;
	AG_Box *bo;
	va_list ap;
	u_int i;

	va_start(ap, fmt);
	vsnprintf(text, sizeof(text), fmt, ap);
	va_end(ap);

	win = AG_WindowNew(AG_WINDOW_MODAL|AG_WINDOW_NORESIZE|
	    AG_WINDOW_NOTITLE);
	AG_WindowSetPosition(win, AG_WINDOW_CENTER, 0);
	AG_WindowSetSpacing(win, 8);

	AG_LabelStatic(win, text);

	bo = AG_BoxNew(win, AG_BOX_HORIZ, AG_BOX_HOMOGENOUS|AG_BOX_WFILL);
	for (i = 0; i < nbOpts; i++) {
		bOpts[i] = AG_ButtonNew(bo, "XXXXXXXXXXX");
	}
	AG_WindowShow(win);
	return (win);
}

/* Prompt the user for a floating-point value. */
void
AG_TextPromptFloat(double *fp, double min, double max, const char *unit,
    const char *format, ...)
{
	char msg[AG_LABEL_MAX];
	AG_Window *win;
	AG_VBox *vb;
	va_list args;
	AG_FSpinbutton *fsb;

	va_start(args, format);
	vsnprintf(msg, sizeof(msg), format, args);
	va_end(args);

	win = AG_WindowNew(AG_WINDOW_MODAL|AG_WINDOW_NOVRESIZE);
	AG_WindowSetCaption(win, "%s", _("Enter real number"));
	AG_WindowSetPosition(win, AG_WINDOW_CENTER, 1);

	vb = AG_VBoxNew(win, AG_VBOX_WFILL);
	AG_LabelNew(vb, AG_LABEL_STATIC, msg);
	
	vb = AG_VBoxNew(win, AG_VBOX_WFILL);
	{
		fsb = AG_FSpinbuttonNew(vb, unit, _("Number: "));
		AGWIDGET(fsb)->flags |= AG_WIDGET_WFILL;
		AG_WidgetBind(fsb, "value", AG_WIDGET_DOUBLE, fp);
		AG_FSpinbuttonSetRange(fsb, min, max);
		AG_SetEvent(fsb, "fspinbutton-return", AGWINDETACH(win));
	}
	
	vb = AG_VBoxNew(win, AG_VBOX_HOMOGENOUS|AG_VBOX_WFILL|AG_VBOX_HFILL);
	AG_ButtonAct(vb, _("Ok"), 0, AGWINDETACH(win));

	/* TODO test type */

	AG_WindowShow(win);
	AG_WidgetFocus(fsb->input);
}

/* Create a dialog to edit a string value. */
void
AG_TextEditString(char **sp, size_t len, const char *msgfmt, ...)
{
	char msg[AG_LABEL_MAX];
	AG_Window *win;
	AG_VBox *vb;
	va_list args;
	AG_Textbox *tb;

	va_start(args, msgfmt);
	vsnprintf(msg, sizeof(msg), msgfmt, args);
	va_end(args);

	win = AG_WindowNew(AG_WINDOW_MODAL|AG_WINDOW_NOVRESIZE);
	AG_WindowSetCaption(win, "%s", _("Edit string"));
	AG_WindowSetPosition(win, AG_WINDOW_CENTER, 1);

	vb = AG_VBoxNew(win, AG_VBOX_WFILL);
	AG_LabelNew(vb, AG_LABEL_STATIC, msg);
	
	vb = AG_VBoxNew(win, AG_VBOX_WFILL);
	{
		tb = AG_TextboxNew(vb, NULL);
		AGWIDGET(tb)->flags |= AG_WIDGET_WFILL;
		AG_WidgetBind(tb, "string", AG_WIDGET_STRING, sp, len);
		AG_SetEvent(tb, "textbox-return", AGWINDETACH(win));
	}

	vb = AG_VBoxNew(win, AG_VBOX_HOMOGENOUS|AG_VBOX_WFILL|AG_VBOX_HFILL);
	AG_ButtonAct(vb, _("Ok"), 0, AGWINDETACH(win));

	AG_WindowShow(win);
	AG_WidgetFocus(tb);
}

/* Prompt the user for a string. */
void
AG_TextPromptString(const char *prompt, void (*ok_fn)(AG_Event *),
    const char *fmt, ...)
{
	AG_Window *win;
	AG_Box *bo;
	va_list args;
	AG_Button *btn;
	AG_Textbox *tb;
	AG_Event *ev;

	win = AG_WindowNew(AG_WINDOW_MODAL|AG_WINDOW_NOVRESIZE|
	    AG_WINDOW_NOTITLE);
	AG_WindowSetPosition(win, AG_WINDOW_CENTER, 0);
	AG_WindowSetSpacing(win, 8);

	bo = AG_BoxNew(win, AG_BOX_VERT, AG_BOX_WFILL);
	AG_LabelNew(bo, AG_LABEL_STATIC, prompt);
	
	bo = AG_BoxNew(win, AG_BOX_VERT, AG_BOX_WFILL);
	{
		tb = AG_TextboxNew(bo, NULL);
		AGWIDGET(tb)->flags |= AG_WIDGET_WFILL;
		AG_WidgetFocus(tb);

		ev = AG_SetEvent(tb, "textbox-return", ok_fn, NULL);
		AG_EVENT_GET_ARGS(ev, fmt)
		AG_EVENT_INS_VAL(ev, AG_EVARG_STRING, "string", s,
		    &tb->string[0]);
		AG_AddEvent(tb, "textbox-return", AGWINDETACH(win));
	}

	bo = AG_BoxNew(win, AG_BOX_HORIZ, AG_BOX_HOMOGENOUS|AG_BOX_WFILL);
	{
		btn = AG_ButtonNew(bo, _("Ok"));
		ev = AG_SetEvent(btn, "button-pushed", ok_fn, NULL);
		AG_EVENT_GET_ARGS(ev, fmt);
		AG_EVENT_INS_VAL(ev, AG_EVARG_STRING, "string", s,
		    &tb->string[0]);
		AG_AddEvent(btn, "button-pushed", AGWINDETACH(win));

		AG_ButtonAct(bo, _("Cancel"), 0, AGWINDETACH(win));
	}

	AG_WindowShow(win);
}

/*
 * Parse a command-line font specification and set the default font.
 * The format is <face>,<size>,<flags>. Acceptable flags include 'b'
 * (bold), 'i' (italic) and 'U' (uppercase).
 */
void
AG_TextParseFontSpec(char *fontspec)
{
	char *s, *c;

	if ((s = strsep(&fontspec, ":,/")) != NULL &&
	    s[0] != '\0') {
		AG_SetString(agConfig, "font.face", s);
	}
	if ((s = strsep(&fontspec, ":,/")) != NULL &&
	    s[0] != '\0') {
		AG_SetInt(agConfig, "font.size", atoi(s));
	}
	if ((s = strsep(&fontspec, ":,/")) != NULL &&
	    s[0] != '\0') {
		u_int flags = 0;

		for (c = &s[0]; *c != '\0'; c++) {
			switch (*c) {
			case 'b': flags |= AG_FONT_BOLD;	break;
			case 'i': flags |= AG_FONT_ITALIC;	break;
			case 'U': flags |= AG_FONT_UPPERCASE;	break;
			}
		}
		AG_SetUint(agConfig, "font.flags", flags);
	}
}
